File: runner-python/runner.py
import importlib.util
import json
import os
import sys

# Load configuration from environment
context_data = json.loads(os.getenv("PIPER_CONTEXT", "{}"))


def write_result(data):
  print("PIPER_RESULT_START")
  print(json.dumps(data))
  print("PIPER_RESULT_END")

def main():
  if len(sys.argv) < 2:
    write_result({"error": "No user script provided as argument"})
    sys.exit(1)

  user_script_path = sys.argv[1]
  module_name = "user_code"

  try:
    spec = importlib.util.spec_from_file_location(
        module_name, user_script_path
    )
    user_module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(user_module)

    if not hasattr(user_module, "handler"):
      raise AttributeError(
          "The script must contain a 'handler(context)' function."
      )

    handler = user_module.handler
  except Exception as e:
    write_result({"error": f"Load Error: {str(e)}"})
    sys.exit(1)

  try:
    output = handler(context_data)
    write_result(output or {})
  except Exception as e:
    write_result({"error": str(e)})
    sys.exit(1)


if __name__ == "__main__":
  main()

File: automation_engine/runner-node/runner.js
const fs = require('fs');
const path = require('path');

const context_data = JSON.parse(process.env.PIPER_CONTEXT || "{}");

function writeResult(data) {
    process.stdout.write("PIPER_RESULT_START\n");
    process.stdout.write(JSON.stringify(data || {}));
    process.stdout.write("\nPIPER_RESULT_END\n");
}

async function main() {
    const userScriptFile = process.argv[2];

    if (!userScriptFile) {
        writeResult({ error: "No user script provided" });
        process.exit(1);
    }

    let handler;
    try {
        const targetPath = path.isAbsolute(userScriptFile)
            ? userScriptFile
            : path.join(process.cwd(), userScriptFile);

        const userModule = require(targetPath);
        handler = userModule.handler;

        if (typeof handler !== 'function') {
            throw new Error("Export a 'handler' function: module.exports = { handler };");
        }
    } catch (e) {
        writeResult({ error: `Load Error: ${e.message}` });
        process.exit(1);
    }

    try {
        const output = await handler(context_data);
        writeResult(output || {});
    } catch (e) {
        writeResult({ error: `Execution Error: ${e.message}` });
        process.exit(1);
    }
}

main();